worlds/Files.py

```diff
@@ -135,31 +130,84 @@ def read(self, file: Optional[Union[str, BinaryIO]] = None) -> None:
                         message = f"{arg0} - "
                 raise InvalidDataError(f"{message}This might be the incorrect world version for this file") from e
 
-    def read_contents(self, opened_zipfile: zipfile.ZipFile) -> None:
+    def read_contents(self, opened_zipfile: zipfile.ZipFile) -> Dict[str, Any]:
         with opened_zipfile.open("archipelago.json", "r") as f:
             manifest = json.load(f)
         if manifest["compatible_version"] > self.version:
             raise Exception(f"File (version: {manifest['compatible_version']}) too new "
                             f"for this handler (version: {self.version})")
+        return manifest
+
+    def get_manifest(self) -> Dict[str, Any]:
+        return {
+            # minimum version of patch system expected for patching to be successful
+            "compatible_version": 5,
+            "version": container_version,
+        }
+
+
+class APWorldContainer(APContainer):
+    """A zipfile containing a world implementation."""
+    game: str | None = None
+    world_version: "Version | None" = None
+    minimum_ap_version: "Version | None" = None
+    maximum_ap_version: "Version | None" = None
+
+    def read_contents(self, opened_zipfile: zipfile.ZipFile) -> Dict[str, Any]:
+        from Utils import tuplize_version, Version
+        manifest = super().read_contents(opened_zipfile)
+        self.game = manifest["game"]
+        for version_key in ("world_version", "minimum_ap_version", "maximum_ap_version"):
+            if version_key in manifest:
+                setattr(self, version_key, Version(*tuplize_version(manifest[version_key])))
+        return manifest
+
+    def get_manifest(self) -> Dict[str, Any]:
+        manifest = super().get_manifest()
+        manifest["game"] = self.game
+        for version_key in ("world_version", "minimum_ap_version", "maximum_ap_version"):
+            version = getattr(self, version_key)
+            if version:
+                manifest[version_key] = version.as_simple_string()
+        return manifest
+
+
+class APPlayerContainer(APContainer):
+    """A zipfile containing at least archipelago.json meant for a player"""
+    game: ClassVar[Optional[str]] = None
+
+    player: Optional[int]
+    player_name: str
+    server: str
+
+    def __init__(self, path: Optional[str] = None, player: Optional[int] = None,
+                 player_name: str = "", server: str = ""):
+        super().__init__(path)
+        self.player = player
+        self.player_name = player_name
+        self.server = server
+
+    def read_contents(self, opened_zipfile: zipfile.ZipFile) -> Dict[str, Any]:
+        manifest = super().read_contents(opened_zipfile)
         self.player = manifest["player"]
         self.server = manifest["server"]
         self.player_name = manifest["player_name"]
+        return manifest
```

Comment:
Isn't modifying self and returning something at the same time a bit ambiguous, especially without doc string explaining that?

Reply:
That's already how patch files work.

Reply:
Can we have a doc string?
`"""Reads a manifest from an opened zipfile, updates the APContainer object and returns the full manifest."""` maybe?

Reply:
> That's already how patch files work.

This looks wrong to me. In the `main` branch right now, none of the `read_contents` return anything.

Reply:
https://github.com/ArchipelagoMW/Archipelago/pull/4331/files#r1921592831
Since this is where it came in, I don't know if it's better to have the discussion there.